src/GafferSceneUI/Inspector.cpp

```diff
@@ -0,0 +1,352 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2021, Cinesite VFX Ltd. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "GafferSceneUI/Private/Inspector.h"
+
+#include "GafferScene/TweakPlug.h"
+
+#include "Gaffer/ScriptNode.h"
+#include "Gaffer/Spreadsheet.h"
+
+#include "boost/bind.hpp"
+
+using namespace IECore;
+using namespace Gaffer;
+using namespace GafferScene;
+using namespace GafferSceneUI::Private;
+
+//////////////////////////////////////////////////////////////////////////
+// Internal utilities
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+/// \todo Why would this walk past all the output plugs except the last one? I
+/// suspect this wasn't the intention.
```

Comment:
This was taken verbatim (well with a Python->C++ conversion) from the _SceneViewInspector :

https://github.com/GafferHQ/gaffer/blob/main/python/GafferSceneUI/_SceneViewInspector.py#L894-L904

> it is weird that it can return an output for the first plug

In this case it doesn't really have any choice, but I don't think it will be called with output plugs anyway.

> why not just use plug->source(), and reject it if it isn't an input?

I believe this is mainly motivated by animation. When a plug is animated, `source()` will return the output from the animation node, but we want the user to be editing the animated plug itself. There's some coverage for this in `ParameterInspector.testSourceAndEdits`.

I do think this will benefit from refactoring in the future, but there was a _lot_ of work to do in generalising the SceneViewInspector, and I felt that this particular battle wasn't one to pick right now.